### PR TITLE
ID-1126 Access Token Cache

### PR DESCRIPTION
**Pull request by tlangs**
Jira: https://broadworkbench.atlassian.net/browse/ID-1126

What:

ECM _MUST_ cache access tokens. Getting an access token from fence providers for every request _tanks_ fence provider performance. Glad we caught this in pre-prod testing!

### CODE REVIEW

**Comment by tlangs** [comment]
I also updated the `publish-branch-image` github action so that I could test the changes in Dev before merging. That should make things easier in the future!

File: service/src/main/java/bio/terra/externalcreds/dataAccess/AccessTokenCacheDAO.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+            .addValue("expiresAt", Timestamp.from(accessTokenCacheEntry.getExpiresAt()));
+
+    // generatedKeyHolder will hold the id returned by the query as specified by the RETURNING
+    // clause
+    var generatedKeyHolder = new GeneratedKeyHolder();
+    var numUpdated = jdbcTemplate.update(query, namedParameters, generatedKeyHolder);
+    if (numUpdated == 1) {
+      return accessTokenCacheEntry;
+    } else {
+      throw new RuntimeException("Failed to upsert access token cache entry");
```
Comment (Ghost-in-a-Jar) [inline service/src/main/java/bio/terra/externalcreds/dataAccess/AccessTokenCacheDAO.java]: Do we expect to ever hit this error? Maybe it should be a more specific message?

File: service/src/main/resources/db/changelog/changesets/20240402_add_access_token_cache_table.yaml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+              - column:
+                  name: access_token
+                  type: text
+                  constraints:
+                    nullable: false
+              - column:
+                  name: expires_at
+                  type: timestamp
+                  constraints:
+                    nullable: false
```
Comment (Ghost-in-a-Jar) [inline service/src/main/resources/db/changelog/changesets/20240402_add_access_token_cache_table.yaml:27]: I dont see a primary key?

**Review by Ghost-in-a-Jar — APPROVED** [review]
Some small comments but nothing blocking

File: service/src/main/resources/db/changelog/changesets/20240402_add_access_token_cache_table.yaml (hunk 2 above)
Comment (tlangs) [reply service/src/main/resources/db/changelog/changesets/20240402_add_access_token_cache_table.yaml:27]: The `linked_account_id` serves as a unique key. There doesn't need to be a primary key on this table.

File: service/src/main/java/bio/terra/externalcreds/services/TokenProviderService.java [hunk 3]
```diff
@@ -113,24 +104,6 @@ public Optional<String> getProviderAccessToken(
...
-    // save the linked account with the new refresh token to replace the old one
-    var refreshToken = accessTokenResponse.getRefreshToken();
-    if (refreshToken != null) {
-      linkedAccountService.upsertLinkedAccount(
-          linkedAccount.withRefreshToken(refreshToken.getTokenValue()));
-    }
-
-    logGetProviderAccessToken(linkedAccount, auditLogEventBuilder);
-
-    return Optional.of(accessTokenResponse.getAccessToken().getTokenValue());
```
Comment (tlangs) [inline service/src/main/java/bio/terra/externalcreds/services/TokenProviderService.java:134]: This was all moved to `AccessTokenCacheService` to contain all the access token logic in one place. This way, there only one entrypoint to get an access token, and other services, like fence key retrieval, can use it.

File: service/src/test/java/bio/terra/externalcreds/dataAccess/AccessTokenCacheDAOTest.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+
+class AccessTokenCacheDAOTest extends BaseTest {
+
+  @Autowired private LinkedAccountDAO linkedAccountDAO;
+  @Autowired private AccessTokenCacheDAO accessTokenCacheDAO;
+
+  @Test
+  void testGetMissingAccessTokenCacheEntry() {
```
Comment (Shakespeared) [inline service/src/test/java/bio/terra/externalcreds/dataAccess/AccessTokenCacheDAOTest.java:21]: Nit: it's nice to have the AAA comments in tests, but these aren't too hard to read so not blocking

**Review by Shakespeared — APPROVED** [review]
Seems reasonable!

**Pull request merged**